crates/boot_srv/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+#![deny(missing_docs)]
+//! Kitsune2 boot server is an HTTP REST server for handling bootstrapping
+//! discovery of peer network reachability in p2p applications.
+//!
+//! Despite being in the kitsune2 repo, `boot_srv` and `boot_cli` do not depend
+//! on any Kitsune2 crates. This is to ensure the bootstrapping functionality
+//! is well-defined, self-contained, easily testable in isolation, and
+//! usable for projects that don't choose to make use of Kitsune2 itself.
+//!
+//! That being said, the boot server and client are designed to transfer the
+//! `AgentInfoSigned` data type defined in the `kitsune2_api` crate. Since
+//! the canonical encoding of that type is JSON, we just redefine a subset
+//! of the schema here, and only validate the parts required for bootstrapping.
+//!
+//! For additional details, please see the [spec].
+
+/// This is a documentation module containing the kitsune2_boot spec.
+///
+/// #### 1. Types
+///
+/// - `Base64Agent` - base64UrlNoPad safe encoded string agent id.
+/// - `Base64Space` - base64UrlNoPad safe encoded string space id.
+/// - `Base64Sig` - base64UrlNoPad safe encoded string crypto signature.
+/// - `Json` - string containing json that can be decoded.
+/// - `I64` - string contaning an i64 number.
+///
+/// ```text
+/// AgentInfoSigned = { "agentInfo": Json, "signature": Base64Sig }
+/// AgentInfo = {
+///   "agent": Base64Agent,
+///   "space": Base64Space,
+///   "createdAt": I64,
+///   "expiresAt": I64,
+///   "isTombstone": boolean
+/// }
+/// ```
+///
+/// Any other properties on these objects will be ignored and pass through.
+///
+/// #### 2. REST API
+///
+/// ##### 2.1. In Brief
+///
+/// ```text
+/// ErrResponse = { "error": string }
+/// OkResponse = {}
+/// ListResponse = [ AgentInfoSigned, .. ]
+/// ```
+///
+/// - `PUT /boot/Base64Space/Base64Agent`
```

Comment:
Should we version the url?

Reply:
I'm inclined towards no... we can always add a `/boot2/` if we really needed, but the structure of the agent info itself is extensible, so any changes we need to make we should be constrainable to additional fields within that.

Reply:
I think the request or response types change the url version should update and requests in the old type explicitly transformed into the new type if both are still supported.

Reply:
The main upgrade scenarios I'm considering are things like using new crypto signing algorithms, which can be accomplished with additional fields, usable through the same REST API. If you send a new info to an old server it will error, but a new server will take and respond with both new and old infos no problem.

In my experience, maintaining multiple API versions requires more overhead in terms of maintenance. Making existing APIs work with new functionality is more up-front work, but less overhead overall.

Reply:
I defer to you but I feel like explicit versioning makes clearer code paths, we get the freedom to deprecate fields if needed, and it makes it explicit that the server you're calling is running on a certain version of the spec.

Reply:
Actually I want to make a stronger case for this.

We know the spec will change, and the functionality will change. Our intention is that all changes will be forwards compatible. By using api versioning we:
- make the forwards compatibility explicit in the codebase, rather than relying on folk knowledge or reviewing old versions of the codebase to confirm this.
- make dropping support for old versions explicit in the codebase
- make explicit to the client which version of the spec is supported by the server it is calling
- allow ourselves the freedom to make breaking changes gracefully. Because we can't be certain we will never want any breaking changes

If we *want* to maintain multiple api versions we can, but having versioning doesn't require that. It just means we're explicit in the codebase and to clients that an api version has been deprecated.

Reply:
[Just FYI, it isn't just me that disagrees with versioning apis](https://www.hmeid.com/blog/just-say-no-to-versioning)

Aside from that I don't think there's any need to *start* with the versioning. If we ever do decide there's no way forward except to version the api. We could introduce a `/v2/` at the front of the path OR an `x-api-ver=2` http header, then say if you don't include those, you get the v1 api while it's deprecated, and then say those are required when we delete the deprecation.

I feel pretty strongly about this as well, but it's not a hill I need to die on. If others also prefer versioned APIs, I'm fine doing it.

Reply:
Yeah that makes sense, we can wait and add if needed. Not the hill I want to die on either, and I'm feeling like I got a bit bikesheddy. I see there are valid arguments both ways.